File: Binary Search/koko_eating_bananas.py
"""
https://leetcode.com/problems/koko-eating-bananas/
"""

class Solution: #Time complexity O(nlogmax(n)), Space complexity O(1)
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        def howlong(k):
            count = 0

            for i in range(len(piles)):
                count += math.ceil(piles[i]/k)

            return count 

        l = 1
        r = max(piles)

        res = r

        while l <= r:
            k = (r+l)//2

            if howlong(k) > h:
                l = k + 1
            else:
                r = k - 1
                res = min(res, k)


        return res

File: Binary Search/koko_eating_bananas.js
/*
https://leetcode.com/problems/koko-eating-bananas/description/
*/


class Solution {
    /**
     * @param {number[]} piles
     * @param {number} h
     * @return {number}
     * Time complexity O(log(max(piles)*len(piles)), Space complexity O(1)
     */
    minEatingSpeed(piles, h) {
        function calcTime(k){
            let t = 0
            for(var pile of piles){
                t += Math.ceil(pile/k)
            }
            return t
        }

        let l = 1
        let r = Math.max(...piles)
        var res = r

        while (l <= r){
            let m = Math.floor((l + r)/2)

            if (calcTime(m) > h){
                l = m + 1
            }else{
                res = m
                r = m - 1
            }
        }
        return res

    }
}
